### PR TITLE
DOC: Update release notes for 1.15.0.

### PR DESCRIPTION
**Pull request by charris**
[ci skip]

### CODE REVIEW

**Review by rgommers — APPROVED** [review]

File: doc/release/1.15.0-notes.rst [hunk 1]
```diff
@@ -47,64 +65,68 @@ Deprecations
...
+  In the future, that will be interpreted as ``arr[array([[0, 1], [0, 1]])]``.
 
 * Direct imports from the following modules is deprecated. All testing related
   imports should come from `numpy.testing`.
+
   * `np.testing.utils`
   * `np.testing.decorators`
   * `np.testing.nosetester`
   * `np.testing.noseclasses`
   * `np.core.umath_tests`
```
Comment (mhvk) [inline doc/release/1.15.0-notes.rst]: `umath_tests` cannot be imported from `np.testing`, making the sentence above incorrect. I would explicitly state it has been turned into a private module to indicate that it may change without notice.

File: doc/release/1.15.0-notes.rst [hunk 2]
```diff
@@ -164,25 +187,33 @@ both numpy 1.15 and 1.16. For more information on how to update code to account
 for this future change see "basics/structured arrays/accessing multiple fields"
```
Comment (mhvk) [inline doc/release/1.15.0-notes.rst]: Can this be made into a direct link?

File: doc/release/1.15.0-notes.rst [hunk 3]
```diff
@@ -261,41 +292,41 @@ Dates, times, and timedeltas can now be histogrammed. The bin edges must be
...
 ``MaskedArray.astype`` now is identical to ``ndarray.astype``
 -------------------------------------------------------------
 This means it takes all the same arguments, making more code written for
 ndarray work for masked array too.
 
 Enable AVX2/AVX512 at compile time
--------------------------------------------------------------
+----------------------------------
 Change to simd.inc.src to use AVX2 or AVX512 at compile time. Solving the gap
 that if compile numpy for avx2 (or 512) with -march=native, still get the SSE
```
Comment (mhvk) [inline doc/release/1.15.0-notes.rst]: I don't understand the "Solving the gap..." sentence...

File: doc/release/1.15.0-notes.rst [hunk 4]
```diff
@@ -436,6 +467,3 @@ within an array.
 .. note:: Implementations of ``__array_ufunc__`` should ensure that they can
           handle either ``axis`` or ``axes``.  In future, we may convert
           ``axis`` to ``axes`` before passing it on.
-
```
Comment (mhvk) [inline doc/release/1.15.0-notes.rst]: Above, l436 is the result of a bad rebase, and should be deleted... Sorry!

**Review by mhvk — COMMENTED** [review]
Small things mostly.

File: doc/release/1.15.0-notes.rst [hunk 5]
```diff
@@ -125,22 +147,24 @@ types from ``ctypes``.
...
-by removing the interception of an otherwise-normal Exception when ``axis`` was provided to an object
-using the old API.
+This is the documented behavior, but previously the result could be any of
+slice, None, or list.
+
+All downstream users seem to check for the `None` result from
+``flatnotmasked_contiguous`` and replace it with ``[]``.  Those callers will
+continue to work as before.
+
+``np.squeeze`` API expectation of objects that cannot handle an ``axis`` argument
```
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: This doesn't feel grammatically meaningful any more

File: doc/release/1.15.0-notes.rst [hunk 6]
```diff
@@ -261,41 +292,41 @@ Dates, times, and timedeltas can now be histogrammed. The bin edges must be
...
 ``histogram`` "auto" estimator handles limited variance better
-------------------------------------------------------------------------
-No longer does an IQR of 0 result in `n_bins=1`, rather the number of bins
+--------------------------------------------------------------
+No longer does an IQR of 0 result in ``n_bins=1``, rather the number of bins
 chosen is related to the data size in this situation.
 
-``histogram`` and ``histogramdd`` return edges matching the float type of the data
-----------------------------------------------------------------------------------
+``histogram`` and ``histogramdd`` edges match the float type of the data
```
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: `return` was a useful disambiguator from the input edges argument. I think this was more correct before, even if the line was long.

File: doc/release/1.14.5-notes.rst [hunk 7]
```diff
@@ -0,0 +1,30 @@
+==========================
+NumPy 1.14.5 Release Notes
+==========================
```
Comment (eric-wieser) [inline doc/release/1.14.5-notes.rst]: A little surprised to see forward-porting release notes from one branch into the next like this.

I think this would make sense as a logically separate commit.

Comment (charris) [reply doc/release/1.14.5-notes.rst]: I thought about it, but the 1.14.5 notes were so small I figured to slip them in here. The oddity was the timing, I probably should have waited another day for the branch.

File: doc/release/1.15.0-notes.rst [hunk 8]
```diff
@@ -47,64 +65,68 @@ Deprecations
     when called with a string.
 
 * Multidimensional indexing with anything but a tuple is
-  deprecated. This means that code such as ``ind = [slice(None), 0]``,
+  deprecated. This means that code such as ``ind = [slice(None), 0]`` and
   ``arr[[slice(None), 0]]`` should be changed to ``arr[tuple(ind)]``. This is
```
Comment (mattip) [inline doc/release/1.15.0-notes.rst]: Is it clear what `ind` is in this context?

File: doc/release/1.15.0-notes.rst [hunk 9]
```diff
@@ -164,25 +187,33 @@ both numpy 1.15 and 1.16. For more information on how to update code to account
...
+The function ``NpyIter_Close`` has been added and should be called before
+``NpyIter_Deallocate`` to resolve possible writeback-enabled arrays.
+
+New functions ``npy_get_floatstatus_barrier`` and ``npy_clear_floatstatus_barrier``
+-----------------------------------------------------------------------------------
+Functions ``npy_get_floatstatus_barrier`` and ``npy_clear_floatstatus_barrier``
+have been added and should be used in place of the ``npy_get_floatstatus``and
+``npy_clear_status`` functions. Optimizing compilers like GCC 8.1 and Clang
+were rearranging the order of operations when the previous functions were used
+in the ufunc SIMD functions, resulting in the floatstatus flags being ' checked
```
Comment (mattip) [inline doc/release/1.15.0-notes.rst]: Extra ' should be removed

File: doc/release/1.15.0-notes.rst [hunk 10]
```diff
@@ -2,12 +2,28 @@
...
+Cython 0.28.2 or later. The wheels will be linked with OpenBLAS 3.0, which
+should fix some of the problems reported for NumPy 1.14.
+
 
 Highlights
 ==========
 
 * NumPy has switched to pytest for testing.
-
+* A new  `np.printoptions` context manager.
```
Comment (mattip) [inline doc/release/1.15.0-notes.rst]: in order for `np.printoptions` to become a link, it must be spelled as `numpy.printoptions`, which will then be rendered with the full module name `numpy`. There are many of these in this doc. We could do any or all of these:

- add double ` and give up on the link
- for this release convert all `np.something` to `numpy.something`
- fix the sphinx namespaces to allow `np.something`

Should I  try to do the last one for `master`?

File: doc/release/1.15.0-notes.rst (hunk 2 above)
Comment (charris) [reply doc/release/1.15.0-notes.rst]: Not sure where to send the link. There is a section in the 1.14 documentation, incorrectly claiming the change will go into 1.15, but I'm not sure how to access that section. See https://docs.scipy.org/doc/numpy/user/basics.rec.html, but is it safe to reference the docs? OTOH, there is the file `numpy/doc/structured_arrays.py`.

Comment (mhvk) [reply doc/release/1.15.0-notes.rst]: Yes, I meant a link to the docs; I don't think it hurts - if it breaks it is still no worse than it is now. Also, I think `Accessing Multiple Fields` might just work... (I think sphinx allows linking to section titles, but may be wrong...)

File: doc/release/1.15.0-notes.rst (hunk 5 above)
Comment (charris) [reply doc/release/1.15.0-notes.rst]: I find the whole entry confusing and have attempted a rewrite.
```
``np.squeeze`` restores old behavior of objects that cannot handle an ``axis`` argument
---------------------------------------------------------------------------------------
Prior to version ``1.7.0``, `np.squeeze` did not have an ``axis`` argument and
all empty axes were removed by default. The incorporation of an ``axis``
argument made it possible to selectively squeeze single or multiple empty axes,
but the old API expectation was not respected because axes could still be
selectively removed (silent success) from an object expecting all empty axes to
be removed. That silent, selective removal of empty axes for objects expecting
the old behavior has been fixed and the old behavior restored.
```
Better?

File: doc/release/1.15.0-notes.rst (hunk 10 above)
Comment (mattip) [reply doc/release/1.15.0-notes.rst]: badly worded my question. Should I try to make a PR to modify our documentation system so that `np.printoptions` becomes a valid link, as well as `numpy.printoptions`?

File: doc/release/1.15.0-notes.rst (hunk 2 above)
Comment (charris) [reply doc/release/1.15.0-notes.rst]: I think sphinx requires specific section labels.

**Comment by charris** [comment]
Revised.

File: doc/release/1.15.0-notes.rst (hunk 10 above)
Comment (charris) [reply doc/release/1.15.0-notes.rst]: Hmm, that could be useful, I'd say go for it.

File: doc/release/1.15.0-notes.rst [hunk 11]
```diff
@@ -2,109 +2,148 @@
...
+  * `numpy.loads`
+  * `numpy.core.numeric.load`
+  * `numpy.core.numeric.loads`
+  * `numpy.ma.loads`, `numpy.ma.dumps`
+  * `numpy.ma.load`, `numpy.ma.dump` - these functions already failed on
+    python 3 when called with a string.
+
+* Multidimensional indexing with anything but a tuple is deprecated. This means
+  that the index lists in ``ind = [slice(None), 0]`` and ``arr[[slice(None),0]]``
+  should be changed to tuples. That is necessary to avoid ambiguity in
```
Comment (jakirkham) [inline doc/release/1.15.0-notes.rst]: Might be worth showing the expected syntax here instead of saying "tuples".

File: doc/release/1.15.0-notes.rst [hunk 12]
```diff
@@ -2,109 +2,148 @@
...
-  * `np.loads`
-  * `np.core.numeric.load`
-  * `np.core.numeric.loads`
-  * `np.ma.loads`, `np.ma.dumps`
-  * `np.ma.load`, `np.ma.dump` - these functions already failed on python 3,
-    when called with a string.
-
-* Multidimensional indexing with anything but a tuple is
-  deprecated. This means that code such as ``ind = [slice(None), 0]``,
-  ``arr[[slice(None), 0]]`` should be changed to ``arr[tuple(ind)]``. This is
```
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: This was supposed to read ``` ``ind = [slice(None), 0]``, ``arr[ind]`` should be changed to ``arr[tuple(ind)]`` ```

File: doc/release/1.15.0-notes.rst [hunk 13]
```diff
@@ -2,109 +2,148 @@
...
+    python 3 when called with a string.
+
+* Multidimensional indexing with anything but a tuple is deprecated. This means
+  that the index lists in ``ind = [slice(None), 0]`` and ``arr[[slice(None),0]]``
+  should be changed to tuples. That is necessary to avoid ambiguity in
+  expressions such as ``arr[[[0, 1], [0, 1]]]``, currently interpreted as
+  ``arr[array([0, 1]), array([0, 1])]``, that will be interpreted
+  as ``arr[array([[0, 1], [0, 1]])]`` in the future.
+
+* Imports from the following sub-modules is deprecated, they will be removed
```
Comment (eric-wieser) [inline doc/release/1.15.0-notes.rst]: *are* deprecated

File: doc/release/1.15.0-notes.rst (hunk 12 above)
Comment (charris) [reply doc/release/1.15.0-notes.rst]: I've taken another cut at this.

File: doc/release/1.15.0-notes.rst [hunk 14]
```diff
@@ -2,109 +2,149 @@
...
+of old functions, and improvements to many existing functions. Please read the
+detailed descriptions below to see if you are affected.
+
+For testing, we have switched to pytest as a replacement for the no longer
+maintained nose framework. The old nose based interface remains for downstream
+projects who may still be using it.
+
+The Python versions supported by this release are 2.7, 3.4-3.6.  The upcoming
+3.7 release should also work, but you will need to compile from source using
+Cython 0.28.2 or later. The wheels will be linked with OpenBLAS 3.0, which
```
Comment (thisch) [inline doc/release/1.15.0-notes.rst]: Could it be that you meant OpenBLAS 0.3 here, which was released in May this year?
